### PR TITLE
[Refactor](push-down predicate) Derive push-down predicate from vconjuncts

### LINKED ISSUE #11467: [Enhancement] Derive push-down predicate from vconjuncts
### Search before asking

- [X] I had searched in the [issues](https://github.com/apache/incubator-doris/issues?q=is%3Aissue) and found no similar issues.

### Description

Currently we always derive  push-down predicates from row-based expressions `conjuncts` although we use vectorized engine.

### Solution

Derive push-down predicates from `vconjuncts` on vectorized engine.

### Are you willing to submit PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### PR DESCRIPTION
**Pull request by Gabriel39**
# Proposed changes

Issue Number: close #11467

## Problem summary

Describe your changes.

## Checklist(Required)

1. Does it affect the original behavior: 
    - [ ] Yes
    - [ ] No
    - [ ] I don't know
2. Has unit tests been added:
    - [ ] Yes
    - [ ] No
    - [ ] No Need
3. Has document been added or modified:
    - [ ] Yes
    - [ ] No
    - [ ] No Need
4. Does it need to update dependencies:
    - [ ] Yes
    - [ ] No
5. Are there any changes that cannot be rolled back:
    - [ ] Yes (If Yes, please explain WHY)
    - [ ] No

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by adonis0147** [comment]
> [Refactor](push-down predicate) WIP

The title of this pr is not good.

**Comment by Gabriel39** [comment]
> > [Refactor](push-down predicate) WIP
> 
> The title of this pr is not good. You can mark the pr as a draft if you are working on it.

I will change this title when this PR is ready for review

File: build.sh [hunk 1]
```diff
@@ -84,7 +84,7 @@ clean_be() {
 
     # "build.sh --clean" just cleans and exits, however CMAKE_BUILD_DIR is set
     # while building be.
-    CMAKE_BUILD_TYPE=${BUILD_TYPE:-Release}
```
Comment (morningman) [inline build.sh:87]: why change this?

File: be/src/vec/exec/volap_scan_node.cpp [hunk 2]
```diff
@@ -621,133 +621,138 @@ Status VOlapScanNode::_add_blocks(std::vector<Block*>& block) {
...
 
     for (int slot_idx = 0; slot_idx < slots.size(); ++slot_idx) {
         switch (slots[slot_idx]->type().type) {
         case TYPE_TINYINT: {
             ColumnValueRange<TYPE_TINYINT> range(slots[slot_idx]->col_name());
-            normalize_predicate(range, slots[slot_idx]);
+            _id_to_slot_column_value_range[slots[slot_idx]->id()] =
+                    std::pair {slots[slot_idx], range};
             break;
         }
```
Comment (yiguolei) [inline be/src/vec/exec/volap_scan_node.cpp:634]: Could we use a template or macro to simply these code?

File: be/src/vec/exec/volap_scan_node.cpp [hunk 3]
```diff
@@ -621,143 +621,71 @@ Status VOlapScanNode::_add_blocks(std::vector<Block*>& block) {
...
+            if (new_root) {
+                (*_vconjunct_ctx_ptr)->set_root(new_root);
+            } else {
+                (*(_vconjunct_ctx_ptr.get()))->mark_as_stale();
+                _stale_vexpr_ctxs.push_back(std::move(_vconjunct_ctx_ptr));
+                _vconjunct_ctx_ptr.reset(nullptr);
+            }
+        }
+    }
+    for (auto it = _id_to_slot_column_value_range.begin();
```
Comment (BiteTheDDDDt) [inline be/src/vec/exec/volap_scan_node.cpp:678]: better use `for(auto it: _id_to_slot_column_value_range)`

File: be/src/vec/exec/volap_scan_node.cpp [hunk 4]
```diff
@@ -898,62 +765,23 @@ Status VOlapScanNode::build_key_ranges_and_filters() {
 Status VOlapScanNode::start_scan(RuntimeState* state) {
     RETURN_IF_CANCELLED(state);
 
-    VLOG_CRITICAL << "Eval Const Conjuncts";
```
Comment (BiteTheDDDDt) [inline be/src/vec/exec/volap_scan_node.cpp:901]: Is it good idea to remove those comment?

File: be/src/vec/exec/volap_scan_node.cpp [hunk 5]
```diff
@@ -966,152 +794,83 @@ static bool ignore_cast(SlotDescriptor* slot, Expr* expr) {
...
         DateV2Value<DateTimeV2ValueType> datetimev2_value =
                 *reinterpret_cast<DateV2Value<DateTimeV2ValueType>*>(value);
-        if (datetimev2_value.can_cast_to_date_without_loss_accuracy()) {
-            DateV2Value<DateV2ValueType> date_v2;
+        if constexpr (IsFixed) {
+            if (datetimev2_value.can_cast_to_date_without_loss_accuracy()) {
+                DateV2Value<DateV2ValueType> date_v2;
+                date_v2.set_date_uint32(
+                        binary_cast<DateV2Value<DateTimeV2ValueType>, uint64_t>(datetimev2_value) >>
+                        TIME_PART_LENGTH);
```
Comment (BiteTheDDDDt) [inline be/src/vec/exec/volap_scan_node.cpp:840]: `>>` seems strange

File: be/src/vec/exec/volap_scan_node.cpp [hunk 6]
```diff
@@ -966,152 +794,83 @@ static bool ignore_cast(SlotDescriptor* slot, Expr* expr) {
...
+        }
+    } else if constexpr ((PrimitiveType == TYPE_DECIMALV2) || (PrimitiveType == TYPE_CHAR) ||
+                         (PrimitiveType == TYPE_VARCHAR) || (PrimitiveType == TYPE_HLL) ||
+                         (PrimitiveType == TYPE_DATETIMEV2) || (PrimitiveType == TYPE_TINYINT) ||
+                         (PrimitiveType == TYPE_SMALLINT) || (PrimitiveType == TYPE_INT) ||
+                         (PrimitiveType == TYPE_BIGINT) || (PrimitiveType == TYPE_LARGEINT) ||
+                         (PrimitiveType == TYPE_DECIMAL32) || (PrimitiveType == TYPE_DECIMAL64) ||
+                         (PrimitiveType == TYPE_DECIMAL128) || (PrimitiveType == TYPE_STRING) ||
+                         (PrimitiveType == TYPE_BOOLEAN)) {
+        if constexpr (IsFixed) {
```
Comment (BiteTheDDDDt) [inline be/src/vec/exec/volap_scan_node.cpp:863]: `IsFixed` It doesn't seem to enter here

File: be/src/vec/exec/volap_scan_node.cpp [hunk 7]
```diff
@@ -30,12 +30,20 @@
...
+#include "vec/exprs/vbloom_predicate.h"
 #include "vec/exprs/vcompound_pred.h"
 #include "vec/exprs/vexpr.h"
+#include "vec/exprs/vruntimefilter_wrapper.h"
+#include "vec/functions/in.h"
 
 namespace doris::vectorized {
 using doris::operator<<;
 
+#define RETURN_IF_PUSH_DOWN  \
```
Comment (BiteTheDDDDt) [inline be/src/vec/exec/volap_scan_node.cpp:42]: Maybe we can change it to RETURN_IF_PUSH_DOWN(stmt) `if(!*push_down) stmt;`

File: be/src/vec/exec/volap_scan_node.cpp [hunk 8]
```diff
@@ -2019,4 +1342,535 @@ Status VOlapScanNode::get_hints(TabletSharedPtr table, const TPaloScanRange& sca
...
+    RETURN_IF_PUSH_DOWN
+    bool opposite = false;
+    VExpr* fn_expr = expr;
+    if (TExprNodeType::COMPOUND_PRED == expr->node_type() &&
+        expr->fn().name.function_name == "not") {
+        fn_expr = fn_expr->children()[0];
+        opposite = true;
+    }
+
+    std::string str;
```
Comment (BiteTheDDDDt) [inline be/src/vec/exec/volap_scan_node.cpp:1726]: better move declare to inner loop

File: be/src/vec/exec/volap_scan_node.cpp [hunk 9]
```diff
@@ -2019,4 +1342,535 @@ Status VOlapScanNode::get_hints(TabletSharedPtr table, const TPaloScanRange& sca
...
+    };
+    if (TExprNodeType::FUNCTION_CALL == expr->node_type()) {
+        if (_should_push_down_function(reinterpret_cast<VectorizedFnCall*>(expr),
+                                       is_null_checker)) {
+            auto temp_range = ColumnValueRange<T>::create_empty_column_value_range(
+                    slot->type().precision, slot->type().scale);
+            temp_range.set_contain_null(true);
+            range.intersection(temp_range);
+            *push_down = true;
+        } else if (_should_push_down_function(reinterpret_cast<VectorizedFnCall*>(expr),
```
Comment (BiteTheDDDDt) [inline be/src/vec/exec/volap_scan_node.cpp:1648]: Seems we can just retain `temp_range.set_contain_null(true);` in the conditional body

File: be/src/vec/exec/volap_scan_node.cpp (hunk 4 above)
Comment (Gabriel39) [reply be/src/vec/exec/volap_scan_node.cpp:901]: I move this evaluation into `_normalize_predicate` so delete this comment.

**Review by BiteTheDDDDt — COMMENTED** [review]
LGTM

File: be/src/vec/exec/volap_scan_node.cpp [hunk 10]
```diff
@@ -257,19 +266,12 @@ Status VOlapScanNode::open(RuntimeState* state) {
...
 
-            for (auto ctx : expr_context) {
-                ctx->prepare(state, row_desc());
-                ctx->open(state);
-                int index = _conjunct_ctxs.size();
-                _conjunct_ctxs.push_back(ctx);
-                // it's safe to store address from a fix-resized vector
-                _conjunctid_to_runtime_filter_ctxs[index] = &_runtime_filter_ctxs[i];
-            }
+            RETURN_IF_ERROR(_append_rf_into_conjuncts(state, vexprs));
```
Comment (HappenLee) [inline be/src/vec/exec/volap_scan_node.cpp:274]: maybe we should after collect all vexprs from `_runtime_filter_descs` to do `_append_rf_into_conjuncts`

File: be/src/vec/exec/volap_scan_node.cpp [hunk 11]
```diff
@@ -2019,4 +1343,530 @@ Status VOlapScanNode::get_hints(TabletSharedPtr table, const TPaloScanRange& sca
...
+
+bool VOlapScanNode::_should_push_down_function_filter(VectorizedFnCall* fn_call,
+                                                      VExprContext* expr_ctx,
+                                                      std::string* constant_str,
+                                                      doris_udf::FunctionContext** fn_ctx) {
+    if (fn_call->fn().name.function_name != "like") {
+        return false;
+    }
+
+    auto children = fn_call->children();
```
Comment (HappenLee) [inline be/src/vec/exec/volap_scan_node.cpp:1383]: use `const auto&` here use auto will do copy

File: be/src/vec/exec/volap_scan_node.cpp [hunk 12]
```diff
@@ -2019,4 +1343,530 @@ Status VOlapScanNode::get_hints(TabletSharedPtr table, const TPaloScanRange& sca
...
+bool VOlapScanNode::_should_push_down_function(
+        VectorizedFnCall* fn_call, const std::function<bool(const std::string&)>& fn_checker) {
+    return fn_checker(fn_call->fn().name.function_name);
+}
+
+bool VOlapScanNode::_should_push_down_function_filter(VectorizedFnCall* fn_call,
+                                                      VExprContext* expr_ctx,
+                                                      std::string* constant_str,
+                                                      doris_udf::FunctionContext** fn_ctx) {
+    if (fn_call->fn().name.function_name != "like") {
```
Comment (HappenLee) [inline be/src/vec/exec/volap_scan_node.cpp:1379]: Add comment of now only support `like` function. wd do not support `not like` ?

File: be/src/exec/olap_utils.h [hunk 13]
```diff
@@ -215,4 +218,25 @@ inline SQLFilterOp to_olap_filter_type(TExprOpcode::type type, bool opposite) {
...
+        return opposite ? FILTER_LESS : FILTER_LARGER;
+    } else if (function_name == "le") {
+        return opposite ? FILTER_LARGER_OR_EQUAL : FILTER_LESS_OR_EQUAL;
+    } else if (function_name == "ge") {
+        return opposite ? FILTER_LESS_OR_EQUAL : FILTER_LARGER_OR_EQUAL;
+    } else if (function_name == "eq") {
+        return opposite ? FILTER_NOT_IN : FILTER_IN;
+    } else if (function_name == "ne") {
+        return opposite ? FILTER_IN : FILTER_NOT_IN;
+    } else if (function_name == "eq_for_null") {
```
[the remark this answers is not in the document]
Comment (Gabriel39) [reply be/src/exec/olap_utils.h:234]: I just copy this from `SQLFilterOp to_olap_filter_type(TExprOpcode::type type, bool opposite)`. After checking, I noticed this condition will never be true. So I think it's better to delete this condition block

File: be/src/vec/exec/volap_scan_node.cpp [hunk 14]
```diff
@@ -2019,4 +1343,530 @@ Status VOlapScanNode::get_hints(TabletSharedPtr table, const TPaloScanRange& sca
...
+        return false;
+    }
+
+    auto children = fn_call->children();
+    doris_udf::FunctionContext* func_cxt = expr_ctx->fn_context(fn_call->fn_context_index());
+    if (!func_cxt) {
```
Comment (HappenLee) [inline be/src/vec/exec/volap_scan_node.cpp:1385]: in which case the `func_cxt == nullptr`  and `childeren.size() !=2` ?
if should not appear, just do DCHECK

File: be/src/vec/exec/volap_scan_node.cpp [hunk 15]
```diff
@@ -2019,4 +1343,530 @@ Status VOlapScanNode::get_hints(TabletSharedPtr table, const TPaloScanRange& sca
...
+}
+
+Status VOlapScanNode::_append_rf_into_conjuncts(RuntimeState* state, std::vector<VExpr*>& vexprs) {
+    if (!vexprs.empty()) {
+        auto last_expr = _vconjunct_ctx_ptr ? (*_vconjunct_ctx_ptr)->root() : vexprs[0];
+        for (size_t j = _vconjunct_ctx_ptr ? 0 : 1; j < vexprs.size(); j++) {
+            if (_rf_vexpr_set.find(vexprs[j]) != _rf_vexpr_set.end()) {
+                continue;
+            }
+            TExprNode texpr_node;
```
[the remark this answers is not in the document]
Comment (Gabriel39) [reply be/src/vec/exec/volap_scan_node.cpp:1847]: Yep, I will do this when we plan to support late-arrival runtime filters

File: be/src/vec/exec/volap_scan_node.cpp (hunk 12 above)
Comment (Gabriel39) [reply be/src/vec/exec/volap_scan_node.cpp:1379]: `not like` is supported now. This processing is included in `_normalize_function_filters`, e.g. a `COMPOUND_PRED` which is `not` will be processed first

File: be/src/vec/exec/volap_scan_node.cpp [hunk 16]
```diff
@@ -2019,4 +1343,530 @@ Status VOlapScanNode::get_hints(TabletSharedPtr table, const TPaloScanRange& sca
...
+    }
+    for (size_t i = 0; i < children.size(); i++) {
+        if (VExpr::expr_without_cast(children[i])->node_type() != TExprNodeType::SLOT_REF) {
+            // not a slot ref(column)
+            continue;
+        }
+        if (!children[1 - i]->is_constant()) {
+            // only handle constant value
+            return false;
+        } else if (children[1 - i]->type().is_string_type()) {
```
Comment (HappenLee) [inline be/src/vec/exec/volap_scan_node.cpp:1395]: why the `like` the other child is not string?

File: be/src/vec/exec/volap_scan_node.cpp [hunk 17]
```diff
@@ -2019,4 +1342,526 @@ Status VOlapScanNode::get_hints(TabletSharedPtr table, const TPaloScanRange& sca
...
+            continue;
+        }
+        if (!children[1 - i]->is_constant()) {
+            // only handle constant value
+            return false;
+        } else {
+            DCHECK(children[1 - i]->type().is_string_type());
+            if (const ColumnConst* const_column = check_and_get_column<ColumnConst>(
+                        children[1 - i]->get_const_col(expr_ctx)->column_ptr)) {
+                *constant_str = std::string(const_column->get_data_at(0).data,
```
Comment (HappenLee) [inline be/src/vec/exec/volap_scan_node.cpp:1399]: const_column->get_data_at(0)->to_string()

File: be/src/vec/exec/volap_scan_node.cpp [hunk 18]
```diff
@@ -2019,4 +1342,526 @@ Status VOlapScanNode::get_hints(TabletSharedPtr table, const TPaloScanRange& sca
...
+                RETURN_IF_ERROR(change_value_range<true>(
+                        not_in_range, value, ColumnValueRange<T>::add_fixed_value_range));
+            }
+            iter->next();
+        }
+        effect = true;
+    } else if (TExprNodeType::BINARY_PRED == expr->node_type()) {
+        DCHECK(expr->children().size() == 2);
+
+        auto ne_checker = [&](const std::string& fn_name) { return fn_name == "ne"; };
```
Comment (HappenLee) [inline be/src/vec/exec/volap_scan_node.cpp:1577]: the checker function should do not use `[&]`, use `[]`

File: be/src/vec/exec/volap_scan_node.cpp [hunk 19]
```diff
@@ -2019,4 +1342,526 @@ Status VOlapScanNode::get_hints(TabletSharedPtr table, const TPaloScanRange& sca
...
+}
+
+bool VOlapScanNode::_should_push_down_binary_predicate(
+        VectorizedFnCall* fn_call, VExprContext* expr_ctx, StringRef* constant_val,
+        int* slot_ref_child, const std::function<bool(const std::string&)>& fn_checker) {
+    if (!fn_checker(fn_call->fn().name.function_name)) {
+        return false;
+    }
+
+    auto children = fn_call->children();
```
Comment (HappenLee) [inline be/src/vec/exec/volap_scan_node.cpp:1417]: recheck all place call `fn_call->children()` use `const auto&`

File: be/src/vec/exec/volap_scan_node.cpp [hunk 20]
```diff
@@ -966,152 +794,83 @@ static bool ignore_cast(SlotDescriptor* slot, Expr* expr) {
...
-        return result_pair;
-    }
-
-    if (pred->get_child(child_idx)->type().type != slot->type().type) {
-        if (!ignore_cast(slot, pred->get_child(child_idx))) {
-            // the type of predicate not match the slot's type
-            return result_pair;
+template <bool IsFixed, PrimitiveType PrimitiveType, typename ChangeFixedValueRangeFunc>
+Status VOlapScanNode::change_value_range(ColumnValueRange<PrimitiveType>& temp_range, void* value,
+                                         const ChangeFixedValueRangeFunc& func, std::string fn_name,
```
Comment (HappenLee) [inline be/src/vec/exec/volap_scan_node.cpp:799]: const std::string& fn_name

File: be/src/vec/exec/volap_scan_node.cpp [hunk 21]
```diff
@@ -2019,4 +1342,534 @@ Status VOlapScanNode::get_hints(TabletSharedPtr table, const TPaloScanRange& sca
...
+
+        auto noneq_checker = [](const std::string& fn_name) {
+            return fn_name != "ne" && fn_name != "eq";
+        };
+        StringRef value;
+        int slot_ref_child = -1;
+        if (_should_push_down_binary_predicate(reinterpret_cast<VectorizedFnCall*>(expr), expr_ctx,
+                                               &value, &slot_ref_child, noneq_checker)) {
+            DCHECK(slot_ref_child >= 0);
+            std::string fn_name =
```
Comment (HappenLee) [inline be/src/vec/exec/volap_scan_node.cpp:1662]: const string&

File: be/src/vec/exec/volap_scan_node.cpp [hunk 22]
```diff
@@ -966,152 +794,83 @@ static bool ignore_cast(SlotDescriptor* slot, Expr* expr) {
...
-    }
-
-    if (pred->get_child(child_idx)->type().type != slot->type().type) {
-        if (!ignore_cast(slot, pred->get_child(child_idx))) {
-            // the type of predicate not match the slot's type
-            return result_pair;
+template <bool IsFixed, PrimitiveType PrimitiveType, typename ChangeFixedValueRangeFunc>
+Status VOlapScanNode::change_value_range(ColumnValueRange<PrimitiveType>& temp_range, void* value,
+                                         const ChangeFixedValueRangeFunc& func,
+                                         std::string& fn_name, int slot_ref_child) {
```
Comment (HappenLee) [inline be/src/vec/exec/volap_scan_node.cpp:800]: not change const std::string

File: be/src/exec/olap_utils.h [hunk 23]
```diff
@@ -215,4 +218,23 @@ inline SQLFilterOp to_olap_filter_type(TExprOpcode::type type, bool opposite) {
     return FILTER_IN;
 }
 
+inline SQLFilterOp to_olap_filter_type(std::string& function_name, bool opposite) {
```
Comment (HappenLee) [inline be/src/exec/olap_utils.h:221]: not change const std::string

**Review by HappenLee — DISMISSED** [review]
LGTM

**Review by yiguolei — APPROVED** [review]
LGTM

**Pull request merged**